cpp/test/sg/kpca_test.cu

```diff
@@ -0,0 +1,418 @@
+      prms.algorithm = solver::COV_EIG_DQ;
+    else
+      prms.algorithm = solver::COV_EIG_JACOBI;
+    int* n_components = (int*)malloc(sizeof(int));
+    *n_components     = params.n_components;
+    kpcaFit(handle, data.data(), alphas.data(), lambdas.data(), n_components, prms, stream);
+
+    raft::update_device(
+      eigenvectors.data(), alphas.data(), prms.n_rows * prms.n_components, stream);
+    raft::update_device(eigenvalues.data(), lambdas.data(), prms.n_components, stream);
+    handle.sync_stream(stream);
+    kpcaTransform(handle,
+                  data.data(),
+                  data.data(),
+                  eigenvectors.data(),
+                  eigenvalues.data(),
+                  trans_data.data(),
+                  prms,
+                  stream);
+  }
+
+  void fitTransformTest()
+  {
+    raft::update_device(data.data(), params.data_h.data(), params.n_rows * params.n_cols, stream);
+    raft::update_device(trans_data_ref.data(),
+                        params.trans_data_ref_h.data(),
+                        params.n_rows * params.n_components,
+                        stream);
+    raft::update_device(
+      eigenvalues_ref.data(), params.eigenvalues_ref_h.data(), params.n_components, stream);
+    raft::update_device(eigenvectors_ref.data(),
+                        params.eigenvectors_ref_h.data(),
+                        params.n_rows * params.n_components,
+                        stream);
+
+    paramsKPCA prms;
+    prms.n_rows             = params.n_rows;
+    prms.n_training_samples = params.n_rows;
+    prms.n_cols             = params.n_cols;
+    prms.n_components       = params.n_components;
+    prms.kernel             = params.kernel;
+    prms.remove_zero_eig    = false;
+    if (params.algo == 0)
+      prms.algorithm = solver::COV_EIG_DQ;
+    else
+      prms.algorithm = solver::COV_EIG_JACOBI;
+
+    int* n_components = (int*)malloc(sizeof(int));
+    *n_components     = prms.n_components;
+    kpcaFit(handle, data.data(), alphas.data(), lambdas.data(), n_components, prms, stream);
+
+    raft::update_device(
+      eigenvectors.data(), alphas.data(), prms.n_rows * prms.n_components, stream);
+    raft::update_device(eigenvalues.data(), lambdas.data(), prms.n_components, stream);
+    handle.sync_stream(stream);
+    kpcaTransformWithFitData(
+      handle, eigenvectors.data(), eigenvalues.data(), trans_data.data(), prms, stream);
+  }
+
+ protected:
+  KPcaInputs<T> params;
+
+  rmm::device_uvector<T> data, trans_data, eigenvectors, eigenvalues, trans_data_ref,
+    eigenvectors_ref, eigenvalues_ref, lambdas, alphas;
+
+  raft::handle_t handle;
+  cudaStream_t stream;
+};
+
+float tolerance           = 0.01f;
+int n_rows                = 3;
+int n_cols                = 2;
+int n_components          = 2;
+int algo                  = 1;
+std::vector<float> data_h = {1.0, 2.0, 5.0, 4.0, 2.0, 1.0};
+
+raft::distance::kernels::KernelParams lin_kern = {raft::distance::kernels::LINEAR, 0, 0, 0};
```

Comment:
Instead of hardcoding the values here, I think we can use our existing PCA as a source of ground truth and validate our results match when we compute the kernel gramm + PCA. That would at least remove a level of hardcoding. This is really hard to maintain. Ideally we'll eventually update our existing PCA tests to compare against a naive PCA solver (like simple manual power iteration) so that we can remove the hardcoding there too.

Reply:
Do you mean doing something like this?
```
import numpy as np
from sklearn.decomposition import PCA, KernelPCA
from sklearn.metrics.pairwise import rbf_kernel

X = np.array([
    [1.0, 4.0],
    [2.0, 2.0],
    [5.0, 1.0],
    [0.0, 3.0],
    [3.0, 5.0],
    [6.0, 2.0],
    [7.0, 8.0],
    [8.0, 1.0],
    [9.0, 6.0],
    [4.0, 7.0]
])

gamma = 0.5
K = rbf_kernel(X, gamma=gamma)

N = K.shape[0]
one_n = np.ones((N, N)) / N
K_centered = K - one_n @ K - K @ one_n + one_n @ K @ one_n

pca = PCA(n_components=2)
X_pca = pca.fit_transform(K_centered)

kpca = KernelPCA(kernel='rbf', gamma=gamma, n_components=2)
X_kpca = kpca.fit_transform(X)

# Results
print("PCA Transformed Data:", X_pca, "KPCA Transformed Data:", X_kpca, sep="\n",end="\n\n")
print("PCA explained_variance_:", pca.explained_variance_, "KPCA eigenvalues_:", kpca.eigenvalues_, sep="\n", end="\n\n")
print("PCA components_:", pca.components_, "KPCA eigenvectors_:", kpca.eigenvectors_, sep="\n", end="\n\n")
```
Looking at the results I see that transformed data and eigenvalues aren't the same between pca/kpca. I haven't had the chance to dig deeper, but I'm wondering if there could be differences in the two algorithms that makes this approach not feasible.

As an alternative, I could rewrite the tests, so we can remove about a lot of the duplicate code for better readability.

Maybe another way to make the test cleaner would be to read the input and expected output from a file. The file could be the output from a Python script using sklearn.KernelPCA.